### ISSUE TITLE
tail call elimination

### ISSUE THREAD

**Issue by gitfoxi** [JuliaLang/julia#4964]
It's interesting that this is valid syntax in Julia and Lua:

```jl
function rec()
    return rec()
end
```

The difference is that in Lua, when you call `rec()` it will stare at you and engage your motherboard's built-in space heater until you ctrl-C out. In Julia:

```jl
julia> rec()
ERROR: stack overflow
 in rec at none:2 (repeats 80000 times)
```

So the stack 80000 things deep. That's interesting.

Why does this matter? I'm not sure. But some people care a lot:

http://www.lua.org/pil/6.3.html

**Comment by oxinabox** [comment]
Recurssion is a very in-style thing for Julia.
Because using multiple dispatch to recognise terminating condition based on change of argument.
Thus programmer are already thinking recussively.

Though I'm not sure the best usecases are solved with tail call recussion optimistation, vs more significant optimisation to flatten a broader class of recussive algorithms.

**Comment by ivandavid77** [comment]
Please consider TCO in implementation of finite machine state, in Erlang is relative easy because this, would be great if Julia include too:

<pre>
<code>
%%%% get_content_service.erl
%% Implement next state machine:
%%
%%   From (A) to (B)  when "begin_tag"
%%   From (B) to (B)  when "content"
%%   From (B) to (A)  when "end_tag"
%%   From (A) to end  when "end_document"
%%
%%                end
%%                 ^                                         ___  content
%%   end_document  |                                       /     \
%%                 |             begin_tag                 |     /
%%             -> (A) ---------------------------------> (B) <-
%%            |                                           |
%%             \ ________________________________________/
%%                                end_tag
%%
%%

-module(get_content_service).
-export([start/0]).
-export([a/1]).


start() ->
  register(get_content_service, spawn(get_content_service, a, [[]])),
  run_test(),
  ok.


a(Values) ->
  receive
    {begin_tag} -> b(Values);
    {end_document, From} -> From ! {result, Values}
  end.


b(Values) ->
  receive
    {content, Value} ->
      NewValues = lists:append(Values, [Value]),
      b(NewValues);
    {end_tag} -> a(Values)
  end.


run_test() ->
  get_content_service ! {begin_tag},
  get_content_service ! {content, 1},
  get_content_service ! {content, 2},
  get_content_service ! {end_tag},
  get_content_service ! {begin_tag},
  get_content_service ! {content, 99},
  get_content_service ! {end_tag},
  get_content_service ! {end_document, self()},
  receive
    {result, Values} -> io:format("~w~n", [Values])
  end.
</code>
</pre>

Then in erlang shell:
<pre><code>
1> c("get_content_service").
{ok,get_content_service}
2> get_content_service:start().
[1,2,99]
ok
</code></pre>

**Comment by carnaval** [comment]
No one here is against TCO, the problem with saying "we have TCO" is that we then have to guarantee that the optimization is performed for any call in tail position, and LLVM does not at the moment provide a way to enforce this. I would expect it to happen in simple cases though.

**Comment by yuyichao** [comment]
I would expect the pop of the GC frame would prevent TCO from happening in many cases... (not this one though)

**Comment by sighoya** [comment]
Correct me if I'am wrong, but can a tail recursive call not be rewritten to a loop by the julia compiler?. I mean that julia recursive code is transformed in loop llvm IR code.

**Comment by yuyichao** [comment]
As an optimization, sure, with very low priority.
As a semantic guarantee, very unlikely.

**Comment by sighoya** [comment]
@yuyichao 
What do you mean with semantic guarantee?

**Comment by yuyichao** [comment]
That functions like

```julia
function rec()
    return rec()
end
```

will be guaranteed to have tail call and not stack overflow.

**Comment by sighoya** [comment]
Thanks for replying.
And why could not the julia compiler write this to
`while(true) end`

**Comment by yuyichao** [comment]
There's no reason it can't for this case, but it most likely can't guarantee that.
Doing that in the compiler level requires guaranteed inferablility (the compiler has to know that you are calling the same function/calling with the same argument types), which is not part of the semantics/guarantee. Doing that without the compiler transformation requires a runtime interface that can do tail call, that's pretty hard/impossible to do (when you need to interleave compiled and "interpreted" frame and somehow tail call both of them).

**Comment by JeffBezanson** [comment]
Local recursion is the easy case. Full tail-call semantics mean that *every* call in tail position must use no stack space, no matter how many functions are involved or what the structure of the call graph is.

**Comment by sighoya** [comment]
Is it a impossible task to guarantee it or does it only slow down your compilation step. If only the latter is true, then something like tag @\doTCO in source code or julia -O3 for the command line would be a solution.
I have another question, is it possible to manipulate the julia compiler via a plugin framework, I've heard that some languages do support it?

**Comment by Keno** [comment]
It is impossible with the current compiler infrastructure if only because LLVM doesn't support it. Whether there is a hypothetical implementation of the language that could provide the guarantee is up in the air. Hooking into the compiler is generally done via generated functions at this point.

**Comment by JeffBezanson** [comment]
It's not impossible, it's just very difficult for little practical benefit. Also, the point of "guaranteed" TCO is not raw performance --- it's not always faster --- the point is to be able to express constant-space control flow with function calls.

Julia supports macros and `@generated` functions, which effectively let you extend the language at two stages. Look through the metaprogramming section of the manual. Beyond that I'd recommend asking more specific questions on discourse, explaining what you'd like to accomplish.

**Comment by pygy** [comment]
Maybe it's been obvious for you all along, but it just occurred to me that you could use more than one LLVM calling convention, @JeffBezanson maybe that's what you meant by "very difficult" rather than "impossible with the current compiler infrastructure".

Couldn't you use a calling convention that supports TCO for Julia => Julia calls, and use CCC for Julia => C and C => Julia? (the latter with a thin wrapper on the C side of things to bridge the CC mismatch... alternatively, Julia functions that are called from both C and Julia could be compiled twice, once for each CC...)

That would enable TCO for pure Julia code, though off course inserting a C frame would break the optimization.

A quick DDG search yielded this: http://lists.llvm.org/pipermail/llvm-dev/2008-December/018955.html so it is not impossible to achieve.

**Comment by yuyichao** [comment]
Using multiple LLVM calling conventions is not the problem,

1. We already don't use C calling convention between julia functions though that does not help tail call at all.
2. The generic case of tail call basically can't avoid calling C function, since type inference and call site de-virtualization (both of which are necessary for removing C calls) are both optimizations and not guaranteed.
3. A different calling convention can't trivially solve this. Ref my comment right above https://github.com/JuliaLang/julia/issues/4964#issuecomment-117299850

Basically to have this as a guarantee rather than an optimization, you'll need to change many different part of the runtime so that they will not get in the way of tail call.

**Comment by pygy** [comment]
Ooh, ok my bad...

Couldn't you let users request TCO explicitly (`@TCO return foo()`?) and have the compiler error out if type inference and devirtualization can't be achieved?

**Comment by yuyichao** [comment]
A feature that depends on type inference and compilation is not something we want to have. We could of course have a very limited version that are defined to work, but it seems hard at a glance to come up with a restriction that's both useful and easily implementable.

**Comment by csaltos** [comment]
Sorry to bother, but after 3 years I wonder if there is tail calls support for Julia now ?

**Comment by PallHaraldsson** [comment]
No, it's easy to check the function at the top, in the latest master, and in my 1-day old master I got:

```
julia> rec()
ERROR: StackOverflowError:
Stacktrace:
 [1] rec()
   @ Main ./REPL[1]:2
```

That said, there's still the different Easter-egg lnaguage (with TCO)::

```
$ julia --lisp
```

**Comment by StefanKarpinski** [comment]
Why do you need tail call elimination?

**Comment by bon** [comment]
> Why do you need tail call elimination?

So that (mutually) recursive functions can be compiled efficiently.

**Comment by PallHaraldsson** [comment]
```
@time fib(big(52_325)) # highest I can do then 1 higher it hangs; on 32-bit Julia I can actually go higher 130_000, not 140_000, so seems like a bug.
```

Isn't the/a `@tailrec` macro good enough (it doesn't support mutual recursion), which gets around the limitation?

```
julia> @time fib(big(1000_000));
 11.860961 seconds (5.00 M allocations: 40.453 GiB, 6.45% gc time)

julia> @time fib(big(100_000));
  0.209107 seconds (500.01 k allocations: 418.178 MiB, 10.54% gc time)

julia> @time fib(100_000);
  0.000259 seconds

julia> @btime fib(100_000);
  56.294 μs (0 allocations: 0 bytes)


(@v1.8) pkg> add https://github.com/TakekazuKATO/TailRec.jl

julia> using TailRec
julia> @tailrec fib(n, a=zero(n), b=one(n)) = n == 0 ? a : fib(n-1, b, a+b)
fib (generic function with 3 methods)


julia> @benchmark fib(46)  # Without the macro is 51x slower, that's the largest Int that doesn't overflow, and maybe an argument to have the macro in Base
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.957 ns … 28.149 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.040 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.088 ns ±  0.871 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █     █▃   ▂                                               ▂
  █▃▄▁█▁██▆▄▇█▄▆▃█▅▄▄▆▅▁▃▅▄▁▁▄▄▃▃▁▃▄▃▄▃▅▅▅▇▇███████▇███▇▇▆▅▅ █
  1.96 ns      Histogram: log(frequency) by time     2.67 ns <

julia> @benchmark fib(big(50_000))  # 14% faster then without the macro. Without big and the macro 3x slower, but both wrong results, except for modulo math:
BenchmarkTools.Trial: 83 samples with 1 evaluation.
 Range (min … max):  48.710 ms … 164.389 ms  ┊ GC (min … max): 13.67% … 70.29%
 Time  (median):     58.329 ms               ┊ GC (median):    18.29%
 Time  (mean ± σ):   60.706 ms ±  14.401 ms  ┊ GC (mean ± σ):  20.66% ±  8.19%

     ▅▂█▂ ▃  ▂▃                                                 
  ▃▃█████▅██▆██▃█▅▁▃▃▃▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃ ▁
  48.7 ms         Histogram: frequency by time          121 ms <

 Memory estimate: 107.83 MiB, allocs estimate: 250006.
```

**Comment by LilithHafner** [comment]
@PallHaraldsson `@tailrec` is seriously outperforming the loop I wrote. Do you know how this is happening?
```jl
#]add  https://github.com/TakekazuKATO/TailRec.jl
using TailRec

@tailrec fib_tr(n, a=zero(n), b=one(n)) = n == 0 ? a : fib_tr(n-1, b, a+b)
#fib_tr (generic function with 3 methods)

function fib_loop(n::Integer)
           a = zero(n)
           b = one(n)
           for _ in 1:n
               a, b = b, a+b
           end
           a
       end
#fib_loop (generic function with 1 method)

fib_tr(big(10_000)) == fib_loop(big(10_000))
#true

@btime fib_tr(big(10_000));
 # 2.398 ms (50006 allocations: 5.01 MiB)

@btime fib_loop(big(10_000));
#  4.300 ms (100006 allocations: 6.24 MiB)

```

**Comment by PallHaraldsson** [comment]
Yes, they are not strictly equivalent, more like which is only 1% slower:
```
julia> function fib_loop(n::Integer)
                  a = zero(n)
                  b = one(n)
                  for _ in n-1:-1:0  # bit fewer allocations with n:big(-1):big(0)
                      a, b = b, a+b
                  end
                  a
              end
```

I admit I had to stare at the code a bit (after noticing the difference in number of allocations, seems subtracting 1 from BigInt is optimized for allocations; adding 1 can make the the result (for a positive number) larger, by one bit, so it seem like they make that assumption allocating always, while subtracting 1 from a positive number can't).

and then in 58% of the time with even fewer allocations:
```
julia> function fib_loop(n::Integer)
                  a = zero(n)
                  b = one(n)
                  for _ in 0:Int64(n)  # should be safe here?
                      a, b = b, a+b
                  end
                  a
              end
```

**Comment by StefanKarpinski** [comment]
It's been mentioned in several discussions of tail call elimination that it's not really an optimization. Optimizations are optional transforms that do not change the behavior of the program but make it faster. Tail call elimination is not optional when you need it; it changes the behavior of the program—stack overflow versus no stack overflow, and in the case of an error, it changes the stack traces substantially; and it's often not faster to reuse the current stack frame than to start a new one. That's why the title of this issue was changed from "tail call optimization" to "tail call elimination".

All of this suggests that it is probably better to have an explicit tail call annotation—after all, when you need it, the compiler is not really free to not do the "optimization" and if the code changes in such a way that the compiler cannot do the transformation anymore (e.g. you accidentally put some over operation around the tail call), then it should be an error rather than silently working by doing normal recursion. I'm not the only one to observe this: https://news.ycombinator.com/item?id=15698624.

Anyway, I think it would make sense to introduce a `@tailcall f(...)` macro—or maybe `@goto f(...)` or `@return f(...)`—that explicitly replaces the current stack frame with a new function invocation. If you call it with anything besides a function invocation, it should error at macro expansion time. It would also allow invoking any function, not just the one that's currently running, since one of the use cases for tail calls is implementing state machines by transferring between a set of functions without pushing the stack. Another feature of an explicit `@tailcall` is that since it explicitly replaces the current function on the stack frame, if an error occurs, those other frames will never appear in the stack trace.

**Comment by StefanKarpinski** [comment]
I guess an aggressive alternative would be that any time someone does `return f(...)` we consider it a tail call and replace the stack frame rather than recursing normally, but that feels like a bit too much.

**Comment by ron-wolf** [comment]
@StefanKarpinski Hm. Devil's advocate: maybe implicit TCE would be useful as a compiler/interpreter option?

**Comment by StefanKarpinski** [comment]
Why? If you're writing code that needs TCE then you _need_ it and being able to run Julia with/without just means your code might be broken by something you don't control. What is the use case for optional TCE? Explicitly required TCE is the only thing that makes sense to me, which then shouldn't be controlled by a flag, it should be a semantic guarantee.

**Comment by NightMachinary** [comment]
@StefanKarpinski I don’t know about the compiler/interpreter flag, but having some kind of globalish switch (e.g., on the module/package level) is very helpful on brevity. I suggest having the @tailcall macro support modules as well, where all the tail calls in that modules get optimized.

**Comment by StefanKarpinski** [comment]
> If you call it with anything besides a function invocation, it should error at macro expansion time. It would also allow invoking any function, not just the one that's currently running, since one of the use cases for tail calls is implementing state machines by transferring between a set of functions without pushing the stack.

A problem with this is that since basically everything is a method call in Julia, what would you even error on? For example, suppose someone is tweaking the definition of a tail recursive function `f` and changes `@tailcall f(...)` into `@tailcall 2*f(...)`. Well, that can still be turned into a tail call, but the call to `f` will use stack space because it's not in tail position.

**Comment by StefanKarpinski** [comment]
This isn't an issue for the pattern where you explicitly indicate that you want tail recursion since then it's clear that if the tail call isn't to the same function, then it should be an error.

I don't really get the brevity thing. `@return f(...)` is only on character longer than `return f(...)` and `@tailcall f(...)` is only three characters longer. I guess you want implicit return by having `f(...)` be the last expression in the function body also get tail call elimination treatment? This is kind of a niche feature that the language has lived without for a decade, so worrying about brevity seems a bit unnecessary. But if the feature exists in the language, then writing a block macro that turns every implicit return into a tail call is easy, so I think we can dispense what that concern here.

**Comment by PallHaraldsson** [comment]
>it's also not necessarily faster to reuse the current stack frame than to start a new one.

Do you have a counterexample? It seems it always should be faster or could be at least as fast, at *runtime*.

What I have in mind is *only* the local case (we could give a guarantee for it, postpone deciding for other cases) that Jeff wrote about:

>Local recursion is the easy case. Full tail-call semantics mean that every call in tail position must use no stack space, no matter how many functions are involved or what the structure of the call graph is.

If TCE is sometimes slower then we do need both options. I'm not pushing for any change, because I know of the `@tailrec` macro. Most users, or beginners will not, so it's important to know how much slower the non-TCE case could be, at runtime. My worry is that even if it were faster at then, how much slower is it at compile time? Is that, only, what you had in mind with potentially slower?

In at least one Julia style-guide:
https://github.com/invenia/BlueStyle
>When using long-form functions always use the return keyword

Since there are two ways to return (a called function result), without the keyword or not, would it be on the table to only have the TCE guarantee only one way, e.g. without `return`, and only as then as simple function call (e.g. not for the `2*f(...)` case?

As soon as we provide a TCE guarantee, we can't really go back, as it would be a breaking change. And doing it as an optimization is worse, if you have it sometimes, you don't know when and why it went away.

From your link:
>Scala sort of has this with the @tailrec annotation. The compiler will always attempt TCO, but the annotation makes failure to do so a compilation error. (Its not a perfect situation though, since there can still be functions assumed to be tail recursive without @tailrec, and whose behavior can change without warning)

I don't think we need a global option, to opt-in to TCE, but we do need one, to opt-out, likely as part of a debug/dev mode (there's a lot more we could have as part of such a mode, a separate discussion). We also need some way to opt out of TCE locally (if it were the default, and assuming sometimes slower), but I'm undecided on if it should be with skipping an explicit `@tailrec`.

**Comment by PallHaraldsson** [comment]
>LLVM does not at the moment provide a way to enforce this. I would expect it to happen in simple cases though.

Is that for sure? I see new in Clang (then implies LLVM supports?):

https://releases.llvm.org/13.0.0/tools/clang/docs/ReleaseNotes.html
>Guaranteed tail calls are now supported with statement attributes [[clang::musttail]] in C++ and __attribute__((musttail)) in C.

Julia can already provide tailrec, as done with `@tailrec` (I'm, not up to speed on all limitations). If LLVM provides support, is that a good enough reason to include support in Julia?

**Comment by FelixBenning** [comment]
The following might be an interesting toy example if anyone would be interested in doing this. It was inspired by this question https://discourse.julialang.org/t/iterator-over-upper-triangular-part-of-cartesian-product/63118, where it appeared that `combinations` from `Combinatorics.jl` was apparently very inefficient. So I tried to generalize the "pairs" algorithm proposed in that thread and ended up using recursion:

```julia
function combinations(x,n)
    if n <= 0
    	return ( () )
    end
    if n== 1
    	return x
    end
    if n==2 # this case is unneccessary
        return ( (a,b) for (k,a) in enumerate(x) for b in Iterators.drop(x, k) )
    end
    return ( 
	(a, b...) 
	for (k,a) in enumerate(x) 
		for b in combinations(Iterators.drop(x,k), n-1) 
        )
end
```
With the `n==2` case you get
```julia
@btime all(x-> gcd(x...) ==1, comb( (1,2,3,4,5,6 ), 2 ))
53.330 ns (0 allocations: 0 bytes)
```
If you leave out the `n==2` case you get

```julia
@btime all(x-> gcd(x...) ==1, comb( (1,2,3,4,5,6 ), 2 ))
 742.421 ns (37 allocations: 3.08 KiB)
```

Since `combinations(x, 1)` just returns `x`, this could be optimized into the `n==2` case with tail end optimization I think. But writing this algorithm without recursion is harder than I thought. So I think this might be an interesting toy example

**Comment by StefanKarpinski** [comment]
There is no tail recursion in that example.

**Comment by vtjnash** [comment]
To make that more explicit: tail recursion requires a statement `return combinations(...)` somewhere in your function. You don't have that.

**Comment by bbarker** [comment]
> Julia can already provide tailrec, as done with @tailrec (I'm, not up to speed on all limitations). If LLVM provides support, is that a good enough reason to include support in Julia? [No (other) C/C++ compiler needs to support, only LLVM?]

out of my element here, but i believe it depends on the calling convention being used: https://internals.rust-lang.org/t/all-tail-calls-are-sibling-calls/1696/4

**Comment by FelixBenning** [comment]
Ah right, I think I misunderstood tail recursion.

**Comment by TANRIninResulu** [comment]
for recursive tree search algorithms to have tail call optimization implicitly is nice to have in a language InşALLAH

**Comment by LilithHafner** [comment]
@TANRIninResulu, please stop posting off-topic and religious content.
